litex/soc/software/bios/boot.c

```diff
@@ -307,6 +305,118 @@ const char *filename, char *buffer)
 	return size;
 }
 
+#ifdef ETH_UDP_DYNAMIC_IP
+
+static uint8_t set_ip(const char * ip_address, unsigned int * ip_to_change)
+{
+    uint8_t n = 0;
+    uint8_t k = 0;
+    uint8_t i;
+    uint8_t size = strlen(ip_address); 
+    unsigned int ip_to_set[4];
+    char buf[3];
+
+    if (size < 7 || size > 15){
+    	printf("Error: Invalid IP address length.");
+	return -1;
+    }
+
+    /* Extract numbers from input, check for potential errors */    
+    for (i = 0; i < size; i++){
+    	if ((ip_address[i] == '.' && k != 0) || (ip_address[i] == '\n' && i == size - 1)){
+		ip_to_set[n] = atoi(buf);	
+		n++;
+		k = 0;
+		memset(buf, '\0', sizeof(buf));
+	} else if (ip_address[i] >= '0' && ip_address[i] <= '9' && k < 3) {
+		buf[k] = ip_address[i];
+		k++;	
+	} else {
+        	printf("Error: Invalid IP address format. Correct format is \"X.X.X.X\".");
+		return -1;
+	}
+    }
+    ip_to_set[n] = atoi(buf);    
+    
+    /* Check if a correct number of numbers was extracted from the input*/
+    if (n != 3){
+	   printf("Error: Invalid IP address format. Correct format is \"X.X.X.X\".");
+    	   return -1;
+    }
+   
+    /* Set the extracted IP address as local or remote ip */
+    for (i = 0; i <= n; i++){
+    	ip_to_change[i] = ip_to_set[i];
+    }
+    return 0;
+	
+}
+
+void set_local_ip(const char * ip_address){
+   if (set_ip(ip_address, local_ip) == 0){	
+   	udp_set_ip(IPTOINT(local_ip[0], local_ip[1], local_ip[2], local_ip[3]));
+   	printf("Local IP : %d.%d.%d.%d", local_ip[0], local_ip[1], local_ip[2], local_ip[3]);
+   }
+}
+
+void set_remote_ip(const char * ip_address)
+{
+    if (set_ip(ip_address, remote_ip) == 0){
+	printf("Remote IP : %d.%d.%d.%d", remote_ip[0], remote_ip[1], remote_ip[2], remote_ip[3]);
+    }
+}
+
+void set_mac_addr(const char * mac_address)
```

Comment:
This function seem to do two things: it parses the MAC address from string to a byte array and sets the global mac address array. Perhaps it should be split?

Reply:
I have split the function into parse_mac_addr() and set_mac_addr(), similiarly to the set_local_ip() and parse_ip() functions. The udp_set_mac() function is now in the set_mac_addr() function while the iterating through strings and arrays is in the parse_mac_addr().

I can split it differently if that is not what you meant but then I would also rearrange the ip functions so that group of functions was in one standard.